packages/realm/bindgen/src/realm_js_browser_helpers.h

```diff
@@ -16,18 +16,54 @@ REALM_NOINLINE inline emscripten::val toEmscriptenErrorCode(const std::error_cod
 
     return out;
 }
-REALM_NOINLINE [[noreturn]] inline emscripten::val toEmscriptenException(const std::exception_ptr& e) noexcept
+REALM_NOINLINE inline emscripten::val toEmscriptenException(const std::exception& e) noexcept
+{
+    return emscripten::val::global("Error")(std::string(e.what()));
+}
+REALM_NOINLINE inline emscripten::val toEmscriptenException(const std::exception_ptr& e)
 {
     try {
         std::rethrow_exception(e);
     }
     catch (const std::exception& e) {
-        // TODO throw an actual Error object (maybe using EM_JS)
-        emscripten::val(e.what()).throw_();
+        return toEmscriptenException(e);
     }
     catch (...) {
-        emscripten::val("Unknown Error").throw_();
+        return emscripten::val::global("Error")(std::string("Unknown error"));
     }
 }
+EM_JS(void, copy_buffer_to_typed_view, (emscripten::EM_VAL buffer_handle, emscripten::EM_VAL view_handle), {
+    const buffer = Emval.toValue(buffer_handle);
+    const view = Emval.toValue(view_handle);
+    view.set(new Uint8Array(buffer));
+});
+// Allocate a new C++ buffer big enough to fit the JS buffer
+// Create a JS memory view around the C++ buffer
+// Call TypedArray.prototype.set to efficiently copy the JS buffer into the C++ buffer via the view
+REALM_NOINLINE inline std::string toBinaryData(const emscripten::val array_buffer) noexcept
+{
+    REALM_ASSERT(array_buffer.instanceof(emscripten::val::global("ArrayBuffer")));
+    std::string buf;
```

Comment:
This acts as the C++ buffer, shouldn't this be heap allocated?

Reply:
It eventually gets moved to the string buffers deque in `RealmAddon` - I need to double-check there aren't any extra copies along the way, but the eventual `BinaryData` we construct points to the buffer in the deque, not the stack variable here.